File: src/Other/NextPermutation.java
package Other;

public class NextPermutation {
    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length == 0) return;
        int i = nums.length - 2;
        while(i >= 0 && nums[i] >= nums[i + 1]) i--; // Find 1st id i that breaks descending order
        if(i >= 0) {                           // If not entirely descending
            int j = nums.length - 1;              // Start from the end
            while(nums[j] <= nums[i]) j--;           // Find rightmost first larger id j
            swap(nums, i, j);                     // Switch i and j
        }
        reverse(nums, i + 1, nums.length - 1);       // Reverse the descending sequence
    }

    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public void reverse (int[] nums, int i, int j) {
        while (i < j) {
            swap(nums, i++, j--);
        }
    }

    public static void main(String[] args) {
        NextPermutation n = new NextPermutation();
        int[] input = {4,3,2,1};
        n.nextPermutation(input);
        for (int i = 0; i < input.length; i++) {
            System.out.println(input[i] + ", ");
        }

    }

}

File: py/essential/NextPermutation.py
# http://leetcode.com/problems/next-permutation/description/?envType=problem-list-v2&envId=rabvlt31
# Grind

from typing import List


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return
        n = len(nums)
        i = n - 2
        # Find the first decreasing element from the end
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        if i >= 0:  # If not the entire array in descending order
            # Find the first element that is greater than nums[i]
            j = n - 1
            while j >= 0 and nums[j] <= nums[i]:
                j -= 1
            # Swap the two elements
            nums[i], nums[j] = nums[j], nums[i]
        # Reverse the elements after the first decreasing element
        print(f"[i]: {i}, [j]: {j}")
        self.reverse(nums, i + 1, n - 1)
        # Reverse the elements in the range [start, end]
        # This is a two-pointer approach

    def reverse(self, nums: List[int], start: int, end: int) -> None:
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1


s = Solution()
input = [3, 4, 2, 1]
s.nextPermutation(input)
print(input)
input = [1, 5, 1]
s.nextPermutation(input)
print(input)
input = [1, 3, 2, 4]
s.nextPermutation(input)
print(input)
